File: cpp/227_基本计算器II_medium.cpp
class Solution {
public:
    int calculate(string s) {
        vector<int> stk;
        char preSign = '+';
        int n = s.length();
        int num = 0;
        int i = 0;
        while (i < n) {
            while (i<n && s[i] == ' ') i++;
            while (i<n && s[i]>='0' && s[i]<='9') {
                num = num*10 + (s[i] - '0');
                i++;
            }
            switch (preSign) {
                case '+':
                    stk.push_back(num);
                    break;
                case '-':
                    stk.push_back(-num);
                    break;
                case '*':
                    stk.back() *= num;
                    break;
                default:
                    stk.back() /= num;
                    break;
            }
            while (i<n && s[i] == ' ') i++;
            if (i < n) {
                preSign = s[i];
                i++;
            }
            num = 0;
        }
        return accumulate(stk.begin(), stk.end(), 0);
    }
};

File: python/227_基本计算器II_medium.py
class Solution:
    def calculate(self, s: str) -> int:
        n, stk = len(s), []
        num, i = 0, 0
        preSign = '+'
        while i<n:
            while i<n and s[i]==' ':
                i+=1
            while i<n and s[i]>='0' and s[i]<='9':
                num = num*10 + (ord(s[i])-ord('0'))
                i+=1
            if preSign == '+':
                stk.append(num)
            elif preSign == '-':
                stk.append(-num)
            elif preSign == '*':
                stk[-1] *= num
            elif preSign == '/':
                stk[-1] = abs(stk[-1])//num if stk[-1]>=0 else -(abs(stk[-1])//num)
            while i<n and s[i]==' ':
                i+=1
            if (i<n):
                preSign = s[i]
                i+=1
            num = 0
        return sum(stk)